### DISCUSSION TITLE
Manifest extensibility enhancements: support additive mutations instead of full override or ignore with ManifestPublishingContext / ManifestPublishingCallbackAnnotation

### DISCUSSION THREAD

**Discussion by tdhatcher** [microsoft/aspire#6974] [Ideas]
## Problem
I'm trying to use the manifest of an Aspire model as input to drive some proprietary CLI tooling to further simplify onboarding for teams. A mono repository in Git is not an option to store all source code for dependencies. So ideally a newcomer to the project would only have to have their dev machine setup with the CLI tool, clone the Aspire project and simply run a CLI command to get up an running.

## Idea
Our CLI's magic will then read the manifest for metadata and handle cloning all Git repositories to a relative location the Aspire project expects. 

This eliminates the need for written documentation that must be maintained about which repositories need to be manually cloned to successfully run our Aspire project and allows us to fully automate that process making onboarding even more "turn key".

The Aspire app model might describe this metadata like

``` csharp
var anotherService = builder
  .AddProject<Projects.AnotherService>("anotherService")
  .WithGitRepository("https://dev.azure.com/mycompany/_git/AnotherService")
  ...
```

At this point the addition of the `WithGitRepository` extension will invoke `WithManifestPublishingCallback`
``` csharp
public static IResourceBuilder<T> WithGitRepository<T>(this IResourceBuilder<T> builder, string repositoryUrl)
    where T : ProjectResource
{
    builder.WithAnnotation(new GitRepositoryAnnotation(repositoryUrl));
    builder.WithManifestPublishingCallback(context => {
        context.Writer.WriteStartObject("repository");
        context.Writer.WriteString("url", repositoryUrl);
        context.Writer.WriteEndObject();
    });
    return builder;
}
```

## A few issues arise around extensibility:

1. The `ManifestPublishCallbackAnnotation` lacks a way to express whether its mutation intends to fully override, be ignored, or just be additive. In my case I wish to make it additive, adding a new object key that's it. 
2. Incomplete context: Adding a `ManifestPublishCallbackAnnotation` using the `WithManifestPublishingCallback` does not provide a reference to the resource or access to the default methods on the `ManifestPublishingContext` to delegate back to. I have no recourse but to rewrite my own so it's also lacking exposure of necessary details. My preference would be to avoid reimplementing any of the default implementation and have to maintain parity as time goes on. This could be made more flexible and not be exclusive as full override or not. https://github.com/dotnet/aspire/blob/7f63747d3cca8617a56cdda5439fd86aecfb43cf/src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs#L88-L130
3. If you have multiple extension method calls using `WithManifestPublishingCallback` all wanting to mutate the manifest in subtle ways on an `IResource` the last one wins.

## Potential areas of enhancement: 
1. Provide a property on the `ManifestPublishingCallbackAnnotation` to indicate whether this should be override or additive?
2. Further enrich the `ManifestPublishingContext` which invokes the callback registered with the `ManifestPublishingCallbackAnnotation` to expose more detail about the current context. 
4. Provide an additional overload to `WithManifestPublishingCallback` extensions in which `IResource` reference to the callback is provided with a signature such as  `Func<ManifestPublishingContext, IResource, Task>`
5. Provide an additional overload to `WithManifestPublishingCallback` so that your callback can invoke this `action` delegate rather than the other way around. The context will still create start and end object, but this provides an option to invert control so the callback chooses to invoke it or not if you want the default (similar to how middleware can chose to halt the request pipeline by not calling `await _next(context)`).
https://github.com/dotnet/aspire/blob/7f63747d3cca8617a56cdda5439fd86aecfb43cf/src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs#L124-L128
6. Perhaps it's simpler to normalize around `ManifestPublishingCallbackAnnotation` having some type of property like `ManifiestPublishingEvents` to follow a flexible eventing pattern which contain numerous `Func<ManifestPublishingContext, IResource, Task>` would be more appropriate to modify behavior at different points of the "writing" processes. Similar to how `OpenIdConnectEvents` work. https://github.com/dotnet/aspnetcore/blob/794124fda4e2de1db0e6893dc0cc031a181ce52a/src/Security/Authentication/OpenIdConnect/src/Events/OpenIdConnectEvents.cs#L9-L72

**Comment by davidfowl** [comment]
This is a big area of investment for the next wave of features. No real details yet but this is issue is good motivation.

Is this a publish time feature you’re building or a runtime one?

**Reply by tdhatcher** [reply]
@davidfowl publish time. My initial approach was to take the simplest to first get something working. I was envisioning a run once or as needed scenario.

I could see the manifest (or a flavor of it) being an automated commit by a build pipeline to keep it in sync and stored alongside aspire host in source control or stored in some other arbitrary location that dev tooling could easily retrieve. 

In my case I find it acceptable and similar dev experience of how a node project has a package.json where a developer with basic skill can be expected to first run `npm install` followd by occassional invocations as new code additions require.

_Unrelated feedback: The company I work at, I have a pretty solid use case for Aspire that has pleny of complexity which covers quite a few areas around an existing customer-facing ecosystem. I took a retrofitting approach with the apsire apphost and made it work without any code modifications to the projects and its all running (including the OTEL observability which I wrapped into a hosting startup assembly). The most complex resource is about 25 lines, mostly environment overrides. Made it work, now making it better and finding ways to improve the experience._

_My current goal is to figure out how to automate all of the manual one time setup into more turnkey setup. And have real world working reference for all the other dev teams to model from._

_The ecosystem requires redis, elasticsearch, mssql server, external Okta auth interactions, certificates, secrets hosted elsewhere, 6 dotnet services in code bases, and a several external dotnet services will not run but depend on, 4 angular clients requiring 2 incompatible versions of node, multi-branded hostnames. Being able to represent setting up a resource in 5-20 lines really beats have to explain and maintain documentation (which drifts)._

**Comment by reiv** [comment]
Adding my +1 here; I'm also interested in this feature, for similar reasons (adding custom annotations for an in-house tool).

With [Publishers now in preview](https://learn.microsoft.com/en-us/dotnet/aspire/whats-new/dotnet-aspire-9.2#-publishers-preview) I assume it's possible to extract all of this information and more, as long as you're willing to write your own. But IMO it'd still be convenient to have this inside the manifest.json, as it's language-agnostic.

On that note, and perhaps @davidfowl can answer this one: are there plans to deprecate the manifest format in favor of Publishers   (somewhat implied by the wording of the linked article), or will it remain a part of the ecosystem?

**Reply by davidfowl** [reply]
Absolutely that is the plan. We don’t plan to break what exists today but with publishers giving people way more control, trying to evolve this system intermediate format is no longer a priority.
